[clang] Implement __builtin_stdc_rotate_left, __builtin_stdc_rotate_right

chaitanyav opened the pull request:
This patch adds type-generic rotate builtins that accept any unsigned integer
type. These builtins provide:

- Support for all unsigned integer types, including _BitInt
- Constexpr evaluation capability
- Automatic normalization of rotation counts modulo the bit-width
- Proper handling of negative rotation counts (converted to equivalent
  positive rotations in the opposite direction)
- Implicit conversion support for both arguments for
  types with conversion operators. 

The builtins follow C23 naming conventions.

Resolves https://github.com/llvm/llvm-project/issues/122819

chaitanyav committed:
[clang] Implement __builtin_stdc_rotate_left and __builtin_stdc_rotate_right

This patch adds type-generic rotate builtins that accept any unsigned integer
type. These builtins provide:

- Support for all unsigned integer types, including _BitInt
- Constexpr evaluation capability
- Automatic normalization of rotation counts modulo the bit-width
- Proper handling of negative rotation counts (converted to equivalent
  positive rotations in the opposite direction)
- Implicit conversion support for both arguments, allowing bool, float, and
  types with conversion operators (not limited to record types)

The builtins follow C23 naming conventions.

Resolves #122819

philnik777 commented on clang/test/Sema/builtin-stdc-rotate.c:
```diff
@@ -0,0 +1,20 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s
+// expected-no-diagnostics
+
+char stdc_rotateleft1[__builtin_stdc_rotate_left((unsigned char)0xB1, 3) == 0x8D ? 1 : -1];
```
I'm pretty sure `static_assert`s are the recommended way to test this now.

philnik777 commented on clang/test/Sema/builtin-stdc-rotate.c:
```diff
@@ -0,0 +1,20 @@
+// RUN: %clang_cc1 -fsyntax-only -verify %s
```
Should the target be pinned? Then the `#ifdef`s wouldn't be necessary anymore I think.

philnik777 reviewed (commented):
Oh, you should also update the title.

philnik777 reviewed (approved):
LGTM, but let's wait for one of the Clang people to take another look.

efriedma-quic commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -3642,6 +3642,7 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
   case Builtin::BI__builtin_rotateleft16:
   case Builtin::BI__builtin_rotateleft32:
   case Builtin::BI__builtin_rotateleft64:
+  case Builtin::BI__builtin_stdc_rotate_left:
```
It looks like emitRotate doesn't handle non-power-of-two rotates correctly.  In particular, if the rotate amount is truncated, you need to reduce it with a modulo operation.

efriedma-quic commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3679,6 +3679,35 @@ the arguments. Both arguments and the result have the bitwidth specified
+``__builtin_stdc_rotate_right`` rotates bits to the right. These builtins
+accept any unsigned integer type, including ``_BitInt`` types. The rotation
+count is taken modulo the bit-width of the value being rotated. These builtins
+can be used within constant expressions.
```
What happens if the count is negative?

chaitanyav replied:
understood the behaviour when the count is negative ([unsigned operations with negative numbers](https://lists.llvm.org/pipermail/llvm-dev/2017-July/115975.html#:~:text=Because%20negative%20numbers%20start%20with,signed%20values%20follow%202's%20complement.%20( ))

efriedma-quic commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -2494,6 +2494,12 @@ RValue CodeGenFunction::emitRotate(const CallExpr *E, bool IsRotateRight) {
+  llvm::Value *BitWidthVal = ConstantInt::get(Ty, BitWidth);
+  llvm::Value *SignedRem = Builder.CreateSRem(ShiftAmt, BitWidthVal);
+  llvm::Value *WrappedShift = Builder.CreateAdd(SignedRem, BitWidthVal);
+  ShiftAmt = Builder.CreateURem(WrappedShift, BitWidthVal);
```
You can't recover the information after the shift amount is truncated.

Missing test coverage.

chaitanyav commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3680,6 +3680,37 @@ the arguments. Both arguments and the result have the bitwidth specified
+accept any unsigned integer type, including ``_BitInt`` types. The rotation
+count is treated as an unsigned value and taken modulo the bit-width of the
+value being rotated. Negative rotation counts are converted to their unsigned
+equivalent before the modulo operation. These builtins can be used within
```
please let me update that, i am normalizing the negative numbers to be in [0, bitwidth - 1] range.

philnik777 replied:
Wouldn't it make more sense to simply reject signed types?

chaitanyav commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -767,12 +767,24 @@ def RotateLeft : BitInt8_16_32_64BuiltinsTemplate, Builtin {
 
+def RotateRightg : Builtin {
+  let Spellings = ["__builtin_rotaterightg"];
+  let Attributes = [NoThrow, Const, Constexpr, CustomTypeChecking];
```
@philnik777 @tbaederr please review

efriedma-quic commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -2492,12 +2492,34 @@ RValue CodeGenFunction::emitRotate(const CallExpr *E, bool IsRotateRight) {
+  }
+
+  llvm::Type *ShiftTy = ShiftAmt->getType();
+  llvm::Value *BitWidthInShiftTy = ConstantInt::get(ShiftTy, BitWidth);
```
(And probably the same fix is needed for constant evaluation.)

cor3ntin commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3680,6 +3680,44 @@ the arguments. Both arguments and the result have the bitwidth specified
+accept any unsigned integer type, including ``_BitInt`` types. The rotation
+count is normalized modulo the bit-width of the value being rotated, with
+negative counts converted to equivalent positive rotations (e.g., rotating left
+by -1 is equivalent to rotating left by BitWidth-1). These builtins can be used
```
```suggestion
by ``-1`` is equivalent to rotating left by ``BitWidth-1``). These builtins can be used
```

cor3ntin commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3680,6 +3680,44 @@ the arguments. Both arguments and the result have the bitwidth specified
+
+  unsigned _BitInt(17) odd_val = 0x1ABCD;
+  unsigned _BitInt(17) rotated_odd = __builtin_stdc_rotate_left(odd_val, 5);
+  unsigned _BitInt(17) neg_rotated_odd = __builtin_stdc_rotate_right(odd_val, -3);
+
+  unsigned _BitInt(128) big_val = 0x123456789ULL;
+  unsigned _BitInt(128) rotated = __builtin_stdc_rotate_left(big_val, 5);
+
```
I don't think we need that many examples

chaitanyav replied:
removed few examples

cor3ntin commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -767,12 +767,24 @@ def RotateLeft : BitInt8_16_32_64BuiltinsTemplate, Builtin {
+  let Attributes = [NoThrow, Const, Constexpr, CustomTypeChecking];
+  let Prototype = "void(...)";
+}
+
```
@pinskia Did these name ship in gcc already? the "stdc" is a bit novel and does not add much

cor3ntin commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14139,23 +14139,17 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
   case Builtin::BI__builtin_rotateright16:
   case Builtin::BI__builtin_rotateright32:
   case Builtin::BI__builtin_rotateright64:
+  case Builtin::BI__builtin_stdc_rotate_left:
+  case Builtin::BI__builtin_stdc_rotate_right:
+  case Builtin::BI_rotl8: // Microsoft variants of rotate left
+  case Builtin::BI_rotl16:
+  case Builtin::BI_rotl:
+  case Builtin::BI_lrotl:
+  case Builtin::BI_rotl64:
```
We are changing the behavior of existing builtins, is that intended?

erichkeane reviewed (commented):
Needs:

- Better commit message (in github top-thing)
- Release note

efriedma-quic commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14166,7 +14160,37 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
+    if (AmtBitWidth > BitWidth) {
+      Divisor = APSInt(llvm::APInt(AmtBitWidth, BitWidth), Amt.isUnsigned());
+    } else {
+      Divisor = APSInt(llvm::APInt(BitWidth, BitWidth), Amt.isUnsigned());
```
Bitwidth fits into BitWidth bits, yes... unless you have `(_BitInt(1))-1`, in which case I think this is flipping the sign.

chaitanyav commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2267,6 +2267,48 @@ static bool BuiltinCountZeroBitsGeneric(Sema &S, CallExpr *TheCall) {
+    return true;
+  }
+
+  ExprResult Arg1Res = S.DefaultLvalueConversion(TheCall->getArg(1));
```
implemented the conversions for both

erichkeane reviewed (commented):
1 question in docs, else this is going to be fine for me.

chaitanyav commented:
@erichkeane, the compiler is issuing a warning. I will use a better example
```
test_doc_example.cpp:4:38: warning: implicit conversion from 'unsigned int' to 'unsigned _BitInt(17)' changes value from 3504544 to 96672
      [-Wconstant-conversion]
    4 |   unsigned _BitInt(17) rotated_odd = __builtin_stdc_rotate_left(0x1ABCD, 5);
      |                        ~~~~~~~~~~~   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

chaitanyav commented:
Please let me know if any changes are needed. Also, I don't have write access. I need help merging this if no changes are needed.

efriedma-quic commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14720,23 +14720,17 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
   case Builtin::BI__builtin_rotateright32:
   case Builtin::BI__builtin_rotateright64:
+  case Builtin::BI__builtin_stdc_rotate_left:
+  case Builtin::BI__builtin_stdc_rotate_right:
```
Per https://discourse.llvm.org/t/updating-expectations-for-the-new-constexpr-engine/88853, please also update clang/lib/AST/ByteCode/InterpBuiltin.cpp .

chaitanyav committed:
Fix rotate builtins as per review

* Make divisor unsigned and unify implementations
* Add implementation to InterpBuiltin.cpp

chaitanyav commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -2492,12 +2492,41 @@ RValue CodeGenFunction::emitRotate(const CallExpr *E, bool IsRotateRight) {
+  if (!llvm::isPowerOf2_32(BitWidth) &&
+      E->getArg(1)->getType()->isSignedIntegerType()) {
+    // converting BitWidth to the type of ShiftAmt
+    llvm::Value *BitWidthInShiftTy = ConstantInt::get(ShiftTy, BitWidth);
```
Made the implementations similar to ExprConstant.cpp.

chaitanyav commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -14747,7 +14741,45 @@ bool IntExprEvaluator::VisitBuiltinCallExpr(const CallExpr *E,
+      if (AmtBitWidth > BitWidth) {
+        Divisor = APSInt(llvm::APInt(AmtBitWidth, BitWidth), Amt.isUnsigned());
+      } else {
+        Divisor = APSInt(llvm::APInt(BitWidth, BitWidth), Amt.isUnsigned());
```
Fixed the issue by making Divisor unsigned.

efriedma-quic commented on clang/test/CodeGen/builtin-rotate.c:
```diff
@@ -1,5 +1,7 @@
 // RUN: %clang_cc1 %s -emit-llvm -o - | FileCheck %s
 
+#include<stdint.h>
```
I think you need to specify `-ffreestanding` to make `#include<stdint.h>` portable.

chaitanyav commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2318,6 +2318,67 @@ static bool BuiltinCountZeroBitsGeneric(Sema &S, CallExpr *TheCall) {
+  QualType Arg0Ty = Arg0->getType();
+  if (!Arg0Ty->isUnsignedIntegerType()) {
+    ExprResult ConvArg0Res = S.PerformImplicitConversion(
+        TheCall->getArg(0), S.Context.UnsignedIntTy, AssignmentAction::Passing);
```
Please advise on how to proceed.

chaitanyav replied:
used PerformContextualImplicitConversion to allow for class/struct types with conversion operators. Other types have to be explicitly casted.

chaitanyav committed:
Address feedback from code review
* Use APInt for Divisor
* Move rotation normalization code into ExprConstShared.h

chaitanyav committed:
Specify -ffreestanding to use the stdint header

chaitanyav committed:
Allow implicit conversion for class/struct types with conversion operators. Explicit conversion must be done for all other types.

efriedma-quic commented on clang/lib/AST/ExprConstant.cpp:
```diff
@@ -19156,6 +19168,42 @@ void HandleComplexComplexDiv(APFloat A, APFloat B, APFloat C, APFloat D,
+    NormAmt = APSInt(APInt(AmtBitWidth, 0), NormAmt.isUnsigned());
+  } else {
+    // Divisor is always unsigned to avoid misinterpreting BitWidth as
+    // negative in small bit widths (e.g., BitWidth=2 would be -2 if signed).
```
Despite this comment, `NormAmt.srem(Divisor)` below interprets the divisor as signed.

Fixing this in the general code seems complicated, so maybe just special-case BitWidth == 2: the normalized rotation amount is either 0 or 1, which corresponds to whether the amount is even or odd.

chaitanyav replied:
@efriedma-quic added the special case for bitwidth=2.

chaitanyav committed:
Fix 2-bit rotation normalization

srem treats both operands as signed. In 2-bit representation
2 is interpretted as -2. Since the rotation amount is 0 or 1
in 2-bit, we can determine by checking if the amount is odd or even

efriedma-quic reviewed (approved):
LGTM

chaitanyav committed:
Move ShiftAmt type conversion outside the if-else so that
it can be applied to all bit widths

chaitanyav committed:
Add missing __builtin_stdc_rotate_left, __builtin_stdc_rotate_right
case labels

chaitanyav commented:
am still simplifying the code.

chaitanyav committed:
* Add byte code interpreter to the test.
* Fixed pushInteger to use getIntWidth(QT) for IntAP types

  clang/lib/AST/ByteCode/IntegralAP.h:79:
  void clang::interp::IntegralAP<false>::copy(const APInt &) [Signed = false]: Assertion `BitWidth == V.getBitWidth()' failed.

efriedma-quic reviewed (approved):
LGTM

dyung commented:
The modifications to the test builtin-rotate.c are failing on a buildbot, can someone take a look or revert if you need time to investigate? https://lab.llvm.org/buildbot/#/builders/154/builds/26659
```
******************** TEST 'Clang :: CodeGen/builtin-rotate.c' FAILED ********************
Exit Code: 2
Command Output (stdout):
--
# RUN: at line 1
/home/tcwg-buildbot/worker/clang-armv8-quick/stage1/bin/clang -cc1 -internal-isystem /home/tcwg-buildbot/worker/clang-armv8-quick/stage1/lib/clang/23/include -nostdsysteminc -ffreestanding /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c -emit-llvm -o - | /home/tcwg-buildbot/worker/clang-armv8-quick/stage1/bin/FileCheck /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c
# executed command: /home/tcwg-buildbot/worker/clang-armv8-quick/stage1/bin/clang -cc1 -internal-isystem /home/tcwg-buildbot/worker/clang-armv8-quick/stage1/lib/clang/23/include -nostdsysteminc -ffreestanding /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c -emit-llvm -o -
# .---command stderr------------
# | /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c:98:61: error: __int128 is not supported on this target
# |    98 |                                    uint64_t u64_2, unsigned __int128 u128,
# |       |                                                             ^
# | /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c:108:21: error: __int128 is not supported on this target
# |   108 |   volatile unsigned __int128 result_u128;
# |       |                     ^
# | /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c:190:62: error: __int128 is not supported on this target
# |   190 |                                     uint64_t u64_2, unsigned __int128 u128,
# |       |                                                              ^
# | /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c:199:21: error: __int128 is not supported on this target
# |   199 |   volatile unsigned __int128 result_u128;
# |       |                     ^
# | 4 errors generated.
# `-----------------------------
# error: command failed with exit status: 1
# executed command: /home/tcwg-buildbot/worker/clang-armv8-quick/stage1/bin/FileCheck /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c
# .---command stderr------------
# | FileCheck error: '<stdin>' is empty.
# | FileCheck command line:  /home/tcwg-buildbot/worker/clang-armv8-quick/stage1/bin/FileCheck /home/tcwg-buildbot/worker/clang-armv8-quick/llvm/clang/test/CodeGen/builtin-rotate.c
# `-----------------------------
# error: command failed with exit status: 2
--
********************
```
